indexer/routing.hpp

```diff
@@ -0,0 +1,190 @@
+    WriteToSink(sink, m_onlyRestrictionCount);
+  }
+
+  template <class Source>
+  void Deserialize(Source & src)
+  {
+    m_version = ReadPrimitiveFromSource<uint16_t>(src);
+    m_reserved = ReadPrimitiveFromSource<uint16_t>(src);
+    m_noRestrictionCount = ReadPrimitiveFromSource<uint32_t>(src);
+    m_onlyRestrictionCount = ReadPrimitiveFromSource<uint32_t>(src);
+  }
+
+  void Reset()
+  {
+    m_version = 0;
+    m_reserved = 0;
+    m_noRestrictionCount = 0;
+    m_onlyRestrictionCount = 0;
+  }
+
+  uint16_t m_version;
+  uint16_t m_reserved;
+  uint32_t m_noRestrictionCount;
+  uint32_t m_onlyRestrictionCount;
+};
+
+static_assert(sizeof(RoutingHeader) == 12, "Wrong header size of routing section.");
+
+class RestrictionSerializer
+{
+public:
+  template <class Sink>
+  static void Serialize(routing::RestrictionVec::const_iterator begin,
+                        routing::RestrictionVec::const_iterator firstOnlyIt,
+                        routing::RestrictionVec::const_iterator end, Sink & sink)
+  {
+    SerializeSingleType(begin, firstOnlyIt, sink);
+    SerializeSingleType(firstOnlyIt, end, sink);
+  }
+
+  template <class Source>
+  static void Deserialize(RoutingHeader const & header, routing::RestrictionVec & restrictions,
+                          Source & src)
+  {
+    DeserializeSingleType(routing::Restriction::Type::No, header.m_noRestrictionCount,
+                          restrictions, src);
+    DeserializeSingleType(routing::Restriction::Type::Only, header.m_onlyRestrictionCount,
+                          restrictions, src);
+  }
+
+private:
+  static uint32_t const kDefaultFeatureId;
+
+  /// \brief Serializes a range of restrictions form |begin| to |end| to |sink|.
+  /// \param begin is an iterator to the first item to serialize.
+  /// \param end is an iterator to the element after the last element to serialize.
+  /// \note All restrictions should have the same type.
+  template <class Sink>
+  static void SerializeSingleType(routing::RestrictionVec::const_iterator begin,
+                                  routing::RestrictionVec::const_iterator end, Sink & sink)
+  {
+    if (begin == end)
+      return;
+
+    routing::Restriction::Type const type = begin->m_type;
+
+    routing::Restriction prevRestriction(type, {} /* links */);
+    prevRestriction.m_featureIds.resize(kSupportedLinkNumber, kDefaultFeatureId);
+    for (auto it = begin; it != end; ++it)
+    {
+      CHECK_EQUAL(type, it->m_type, ());
+
+      routing::Restriction const & restriction = *it;
+      CHECK(restriction.IsValid(), ());
+      CHECK_EQUAL(restriction.m_featureIds.size(), kSupportedLinkNumber,
+                  ("Only", kSupportedLinkNumber, "links restriction are supported."));
+
+      BitWriter<FileWriter> bits(sink);
+      for (size_t i = 0; i < kSupportedLinkNumber; ++i)
+      {
+        uint32_t const delta = bits::ZigZagEncode(static_cast<int32_t>(restriction.m_featureIds[i]) -
```

Comment:
Are you sure that this is a good idea to use zig-zag coding for restrictions? I mean, you highly depends on the fact that the size of m_featureIds is the same for all restrictions, but this may change in future, and zig-zag coding will not work. Moreover, did you try to apply delta encodings not for computations of delta between previous and current restrictions (they seems uncorrelated), but for deltas in the same restriction?  I suggest you to encode the first feature id in the restriction as is, and encode other feature ids in the restriction as deltas from the previous feature id in the same restriction.